### PR TITLE
Refactor searching and reporting results in navto.  

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Local traces were showing a few sources of inefficiency when doing a NavTo search in a large solution (esp. durign the typing phase of writing the query):

1. Cancellation was showing up quite high (sometimes more than 10% of the trace), due to high task fan-out, and cascading cancellation upwards.  This changes us to no cancel by throwing, but just bailing out.  
2. High overhead of async calls during CPU heavy operations (searching).  The main searching loop per file has become completely synchronous.  
3. stalls in searching while reporting results.  Results could only be reported one at a time, and involved a cross process hop (to update the UI immediately).  This has no changed to having the parallel search just push results to a channel (immediately allowing hte searching threads to continue).  A dedicated task reads from the channel and reports it to the host.  That task can send multiple items at a time when available.

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Callback.cs [hunk 1]
```diff
@@ -18,11 +20,13 @@ internal partial class NavigateToItemProvider
     {
         private class NavigateToItemProviderCallback : INavigateToSearchCallback
         {
+            private readonly Solution _solution;
             private readonly INavigateToItemDisplayFactory _displayFactory;
             private readonly INavigateToCallback _callback;
 
-            public NavigateToItemProviderCallback(INavigateToItemDisplayFactory displayFactory, INavigateToCallback callback)
+            public NavigateToItemProviderCallback(Solution solution, INavigateToItemDisplayFactory displayFactory, INavigateToCallback callback)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Callback.cs:27]: part of the ystem used to involve a callback that would pass along a project+item.  The former is redundant given that everyone starts with a solution snapshot, and a navto item always contains a project-id in it anyways.

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Callback.cs [hunk 2]
```diff
@@ -39,9 +43,41 @@ public void Done(bool isFullyLoaded)
                 }
             }
 
-            public Task AddItemAsync(Project project, INavigateToSearchResult result, CancellationToken cancellationToken)
+            public Task AddResultsAsync(ImmutableArray<INavigateToSearchResult> results, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemProvider.Callback.cs:46]: 1. naming become more consistent.  There is AddResultsAsync for when there's an array of INavigateToSearchResult, and AddItemsAsync for when it's an array of RoslynNavigateToItem.
2. These callbacks now take arrays so the server can send a lot at once when it has many items available (if the searchign threads are faster than the sending thread).

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs [hunk 3]
```diff
@@ -125,79 +130,61 @@ private static void ClearCachedData()
...
+                    var index = await GetIndexAsync(storageService, documentKey, cancellationToken).ConfigureAwait(false);
+                    if (index == null)
+                        return;
 
-            await SearchCachedDocumentsInCurrentProcessAsync(
-                storageService, patternName, patternContainer, declaredSymbolInfoKindsSet,
-                onItemFound, lowPriDocs, cancellationToken).ConfigureAwait(false);
+                    ProcessIndex(
+                        documentKey, document: null, patternName, patternContainer, declaredSymbolInfoKindsSet,
+                        index, linkedIndices: null, onItemFound, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs:172]: Importantly, searching each index is pure sync.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs [hunk 4]
```diff
@@ -62,4 +65,91 @@ private static PooledDisposer<PooledHashSet<T>> GetPooledHashSet<T>(ImmutableArr
...
+    /// Main utility for searching across items in a solution.  The actual code to search the item should be provided in
+    /// <paramref name="callback"/>.  Each item in <paramref name="items"/> will be processed using
+    /// <code>Parallel.ForEachAsync</code>, allowing for parallel processing of the items, with a preference towards
+    /// earlier items.
+    /// </summary>
+    private static async Task PerformParallelSearchAsync<T>(
+        IEnumerable<T> items,
+        Func<T, Action<RoslynNavigateToItem>, CancellationToken, ValueTask> callback,
+        Func<ImmutableArray<RoslynNavigateToItem>, Task> onItemsFound,
+        CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs:81]: this is hte core utility that all 3 nav-to search functions end up deferring to (search normal docs.  search generated docs.  search cached docs).

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs [hunk 5]
```diff
@@ -62,4 +67,91 @@ private static PooledDisposer<PooledHashSet<T>> GetPooledHashSet<T>(ImmutableArr
...
+
+        foreach (var item in source)
+        {
+            tasks.Add(Task.Run(async () =>
+            {
+                await body(item, cancellationToken).ConfigureAwait(false);
+            }, cancellationToken));
+        }
+
+        await Task.WhenAll(tasks).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs:154]: poor man's solution.  I consider copying the runtime code and forking into roslyn.  but it was complex enough i didn't want us to main that.  And that code should not run here as this is the searching part, which should run in OOP on VS and on netcore for vscode anyways.   (And it's the same logic as before, so no worse for any other client that was previously using hte old codepaths).

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs [hunk 6]
```diff
@@ -120,84 +118,56 @@ private static void ClearCachedData()
 
         using var _1 = GetPooledHashSet(priorityDocumentKeys, out var priorityDocumentKeysSet);
 
-        // Sort the groups into a high pri group (projects that contain a high-pri doc), and low pri groups (those
-        // that don't).
+        // Sort the groups into a high pri group (projects that contain a high-pri doc), and low pri groups (those that
+        // don't), and process in that order.
         using var _2 = GetPooledHashSet(groups.Where(g => g.Any(priorityDocumentKeysSet.Contains)), out var highPriorityGroups);
         using var _3 = GetPooledHashSet(groups.Where(g => !highPriorityGroups.Contains(g)), out var lowPriorityGroups);
```
Comment (ToddGrun) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs:124]: Don't understand why using hashsets here, why not just:

var highPriorityGroups = groups.Where(g => g.Any(priorityDocumentKeysSet.Contains));
var lowPriorityGroups = groups.Where(g => !highPriorityGroups.Contains(g))

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs [hunk 7]
```diff
@@ -120,84 +118,56 @@ private static void ClearCachedData()
...
-
-        async Task ProcessProjectGroupAsync(IGrouping<ProjectKey, DocumentKey> group)
-        {
-            cancellationToken.ThrowIfCancellationRequested();
-            await Task.Yield().ConfigureAwait(false);
             var project = group.Key;
 
-            // Break the project into high-pri docs and low pri docs.
+            // Break the project into high-pri docs and low pri docs, and process in that order.
             using var _1 = GetPooledHashSet(group.Where(priorityDocumentKeysSet.Contains), out var highPriDocs);
```
Comment (ToddGrun) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs:141]: >using var _1 = GetPooledHashSet [](http://example.com/codeflow?start=12&length=31)

Similar question here

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs:124]: I'll change. It's just how things were done before.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs:124]: ok.  i remember why.  the reason here is that the first check already walked everything (including *into* the groups).  the second line doesn't need to walk into the groups at all.  The low priority groups are determined by just checking them against the high pri groups.  using sets makes the second pass O(1) for each group.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs:141]: this one can be simpler

**Comment by CyrusNajmabadi** [comment]
@ToddGrun for eyes.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs (hunk 7 above)
Comment (ToddGrun) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs]: Why not use your Prioritize method here too?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs]: missed that case. happy to do in followup (unless htere is larger stuff found in this pass) :)

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs [hunk 8]
```diff
@@ -62,4 +67,107 @@ private static PooledDisposer<PooledHashSet<T>> GetPooledHashSet<T>(ImmutableArr
...
+        IEnumerable<T> items,
+        Func<T, Action<RoslynNavigateToItem>, CancellationToken, ValueTask> callback,
+        Func<ImmutableArray<RoslynNavigateToItem>, Task> onItemsFound,
+        CancellationToken cancellationToken)
+    {
+        // Use an unbounded channel to allow the writing work to write as many items as it can find without blocking.
+        // Concurrently, the reading task will grab items when available, and send them over to the host.
+        var channel = Channel.CreateUnbounded<RoslynNavigateToItem>(s_channelOptions);
+
+        await Task.WhenAll(
```
Comment (ToddGrun) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs:103]: >await Task.WhenAll( [](http://example.com/codeflow?start=8&length=19)

This pattern seems useful. Could it be made into a class to hide the channel usage and task yield/whenall details?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs:103]: have done so in the FAR pr.  will pull out of this one in that pr.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs [hunk 9]
```diff
@@ -133,50 +88,38 @@ internal abstract partial class AbstractNavigateToSearchService
...
-                document, declaredSymbolInfo, cancellationToken).ConfigureAwait(false);
-
-            var result = ConvertResult(
-                documentKey, document, declaredSymbolInfo, nameMatches, containerMatches, additionalMatchingProjects);
-            await onResultFound(result).ConfigureAwait(false);
+            if (kinds.Contains(declaredSymbolInfo.Kind) &&
+                nameMatcher.AddMatches(declaredSymbolInfo.Name, ref nameMatches.AsRef()) &&
+                containerMatcher?.AddMatches(declaredSymbolInfo.FullyQualifiedContainerName, ref containerMatches.AsRef()) != false)
+            {
+                if (cancellationToken.IsCancellationRequested)
```
Comment (ToddGrun) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs:105]: >if (cancellationToken.IsCancellationRequested) [](http://example.com/codeflow?start=16&length=46)

nit: seems a bit overkill as this was checked earlier inside the loop

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs:105]: i kept this as addmatches is the expensive step.  so this lets us bail after doing that check.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs [hunk 10]
```diff
@@ -217,28 +160,24 @@ internal abstract partial class AbstractNavigateToSearchService
...
-
             // See if the index for the other file also contains this same info.  If so, merge the results so the
             // user only sees them as a single hit in the UI.
             if (index.DeclaredSymbolInfoSet.Contains(declaredSymbolInfo))
-                result.Add(linkedDocument.Project.Id);
+                result.Add(projectId);
         }
 
-        result.RemoveDuplicates();
         return result.ToImmutableAndClear();
```
Comment (ToddGrun) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs:181]: >return result.ToImmutableAndClear(); [](http://example.com/codeflow?start=8&length=36)

Old code had a RemoveDuplicates call here. Just looking for validation that it wasn't necessary.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs:181]: it wasn't necessary.  this is for getting siblings, which alreadywill only be for unique projects.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.NormalSearch.cs [hunk 11]
```diff
@@ -41,14 +40,21 @@ internal abstract partial class AbstractNavigateToSearchService
...
     {
-        return SearchProjectInCurrentProcessAsync(
-            document.Project, priorityDocuments: [], document, searchPattern, kinds,
-            onItemFound, () => Task.CompletedTask, cancellationToken);
+        var (patternName, patternContainerOpt) = PatternMatcher.GetNameAndContainer(searchPattern);
+        var declaredSymbolInfoKindsSet = new DeclaredSymbolInfoKindSet(kinds);
+
+        var results = new ConcurrentSet<RoslynNavigateToItem>();
+        await SearchSingleDocumentAsync(
+            document, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, t => results.Add(t), cancellationToken).ConfigureAwait(false);
```
Comment (ToddGrun) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.NormalSearch.cs:54]: >t => results.Add(t) [](http://example.com/codeflow?start=84&length=19)

StandardCallbacks<RoslynNavigateToItem>.AddToConcurrentSet?

Comment (ToddGrun) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.NormalSearch.cs:54]: would require passing around the set though

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.NormalSearch.cs:54]: yeah.  don't care  here :)

File: src/Features/LanguageServer/Protocol/Handler/Symbols/WorkspaceSymbolsHandler.cs [hunk 12]
```diff
@@ -70,20 +70,26 @@ private sealed class LSPNavigateToCallback(
...
-                var document = await result.NavigableItem.Document.GetRequiredDocumentAsync(project.Solution, cancellationToken).ConfigureAwait(false);
+                Contract.ThrowIfNull(context.Solution);
 
-                var location = await ProtocolConversions.TextSpanToLocationAsync(
-                    document, result.NavigableItem.SourceSpan, result.NavigableItem.IsStale, context, cancellationToken).ConfigureAwait(false);
-                if (location == null)
-                    return;
+                foreach (var result in results)
+                {
+                    var solution = context.Solution;
```
Comment (ToddGrun) [inline src/Features/LanguageServer/Protocol/Handler/Symbols/WorkspaceSymbolsHandler.cs:79]: >var solution = context.Solution; [](http://example.com/codeflow?start=20&length=32)

nit: could move outside loop

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Symbols/WorkspaceSymbolsHandler.cs:79]: can do in followup.

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Features/LanguageServer/Protocol/Handler/Symbols/WorkspaceSymbolsHandler.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Symbols/WorkspaceSymbolsHandler.cs:79]: 03b0fe7f48a

**Pull request merged**